torch/distributions/lkj_cholesky.py

```diff
@@ -69,6 +69,7 @@ def __init__(self, dim, concentration=1., validate_args=None):
         super(LKJCholesky, self).__init__(batch_shape, event_shape, validate_args)
 
     def expand(self, batch_shape, _instance=None):
+        # This needs a custom .expand() because the dim arg is not in arg_constraints.
```

Comment:
If it is simpler, I suppose we can add `dim` to `arg_constraints` but the scalar constraint on `dim` cannot be checked at the moment.

Reply:
Right, I think there is an implicit assumption on `Constraints` that they are batchable / broadcastable, but this `dim` param is a non-batchable constant.  I think the strictly proper way to pass `dim` would be as an array whose shape we read, e.g.
```py
class LKJCholesky(Distribution):
    arg_constraints = {"dim_prototype": constraints.independent(constraints.real, 1),
                       "concentration": constraints.positive}
    def __init__(self, dim_prototype, concentration, validate_args=None):
        self.dim = dim_prototype.size(-1)
        ...
```
but that seems awkwardly pedantic.

Reply:
So far, we have 3 types of args: tensors, constants (str, int,...), and base distributions. For expanding, I guess we can expand tensors and base distributions, and keep the constants as-is. The tricky thing is we currently don't have a mechanism to declare how arguments stored in a distribution instance. E.g. `dim` may be stored in attributes `.dim` or `._dim`. If we can make a restriction that `dim` is stored as `.dim`, then I guess we can inspect the `__init__` method and pass those constants to the expanded instance (alternatively, we can declare `dim` as a special constraint in `arg_constraints` as @neerajprad suggested, or create a new attribute named `extra_args`,...). I can see a few controversial arguments here.

Reply:
@fehiepsi interesting, I guess we could also consider `Transform`s as constants since they are not expanded. I guess we could add an additional piece of data, say `base_distributions : Frozenset[str]` that would work to automate most cases, with a couple weird cases like `IndependentDistribution` and `MixtureSameFamily`...

Reply:
Yes, I think `Transform` (and e.g. neural networks) can be considered as constants. About IndependentDistribution and MixtureSameFamily, I think we can check the base distribution's batch shape to see if its left-most dimensions match the main distribution's batch_shape. If not, we raise not ImplementedError. Otherwise, we perform expanding:
```python
def expand(self, batch_shape):
    # here base_dist can have other names
    new_base_dist = self.base_dist.expand(batch_shape + self.base_dist.batch_shape[len(self.batch_shape):])
```

For more complicated distributions, I guess we need to use custom `_expand` methods (e.g. to not expand base distributions and improve the performance - like in GaussianHMM).